### ISSUE TITLE
@inject(Element) on view rendered by router-view injects body element

### ISSUE THREAD

**Issue by MeirionHughes** [aurelia/framework#139]
If you inject the Element into a ViewModel, where the View (a route basically) is rendered by the router-view. Rather than get the template element, you get the body element, supplied to the ctor. 

ViewModel: 

``` javascript
import {inject} from 'aurelia-framework';

@inject(Element)
export class About {
  constructor(element) {
    console.log(element);
  }
}
```

View:

``` html
<template>
  <section class="au-animate"></section>
</template>
```

debug: 

![img](http://s29.postimg.org/ttfeekcdj/wrong_bind.png)

**Comment by EisenbergEffect** [comment]
Yes, because the template does not exist at runtime. An html template is not ever part of the real DOM. So, you get the parent container.

**Comment by MeirionHughes** [comment]
shouldn't I be getting <router-view> then rather than body?

**Comment by MeirionHughes** [comment]
For instance: if I do: 
```jade
template
  require(from="./widget")
  section(class="au-animate")
  widget
```
with: 
```jade
template
  div(id="widget-root")

the widget ViewModel will be given the <widget> element. So the children of that element will be directly from my template. However, when its rendered as a route, the viewmodel is given <body> rather than <router-view>...

**Comment by EisenbergEffect** [comment]
Hmmm. Yes, that is correct. I'll re-open this to investigate that. Are you using the latest version of Aurelia?

**Comment by MeirionHughes** [comment]
I believe so: 
```
ok   Up to date - aurelia-framework as github:aurelia/framework@^0.12.0 (0.12.0)
ok   Up to date - aurelia-bootstrapper as github:aurelia/bootstrapper@^0.13.1 (0.13.1)
ok   Up to date - aurelia-animator-css as github:aurelia/animator-css@^0.3.2 (0.3.2)
ok   Up to date - aurelia-validation as github:aurelia/validation@^0.2.5 (0.2.5)
```

**Comment by EisenbergEffect** [comment]
Ok, it might be a bug then. Can you provide me some simple instructions to reproduce it with the basic navigation skeleton app? Is it just a matter of trying to inject the element into the Welcome view model, for example?

**Comment by MeirionHughes** [comment]
yeah pretty much. I can fork if you're busy.

**Comment by EisenbergEffect** [comment]
No need, if that's all I need to do, that's enough to go on.

On Mon, Jun 22, 2015 at 4:06 PM, Meirion Hughes <notifications@github.com>
wrote:

> yeah pretty much. I can fork if you're busy.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/aurelia/framework/issues/139#issuecomment-114115440>.
>



-- 
Rob Eisenberg,
President - Blue Spire
www.durandaljs.com

**Comment by MeirionHughes** [comment]
I guess the problem with this is if you have a template with multiple router-views in it. Do you have an encompassing router-view?

**Comment by EisenbergEffect** [comment]
Apologies...I forgot about this. It's not actually possible. The reason is that the router instantiates classes independent of the views. The router-view may or may not even exist yet, so you can't actually inject Element in this case. In general, you don't want to do that for a view-model. It's meant for custom elements and attributes.

If you really need to get ahold of an element, I suggest you use the `ref` binding to grab the specific element in the view you want.

**Comment by thomas-darling** [comment]
@EisenbergEffect I see this still doesn't work, but I was wondering if this is still impossible to support, or if something might have changed? We have a couple of use cases here which would _really_ benefit from being able to inject the router-view element, and it caught us by surprise that it was not possible.

Is there _any_ way one could get a reference to that element through dependency injection?

**What we're trying to acheive**

Specifically, we inject a service into some of our view models, using a custom lifetime which always registers it in the current child container, thus giving us one service instance per view model (by the way, I was thinking of creating a pull request for that one, as it's really handy). What we would like to do, is to inject the router-view element _into that service_, but for that, we of course need the element to be registered in the child container, which it currently isn't.

The use case here has something to do with tracking of user interactions, and for that we would like the Analytics instance injected into each view model to know the value of an attribute applied to the element representing the view, specifying the name to use when tracking interactions within it - when logging an interaction event, we then walk up the DOM, concatenating those names, thus computing a "path" indicating where in the app the interaction occurred.

So when a view model logs an event using the injected Analytics instance:
`this.analytics.log("next-image", {})`
The logged event will automatically get the computed path, such as:
`{ path: "app/hotels/details/about-section/room-pictures", event: "next-image", args: {} }`

The idea is that a default name can then be specified on the template for each router-view/custom-element:
`<template analytics-name="image-gallery">`
This can then, if needed, be overridden when using the component:
`<image-gallery analytics-name="room-pictures">`

I just now realized that attributes applied to the `<template>` for a router view are not copied to the actual router-view element, as is the case for custom element - is that intentional, or a bug?

**Comment by thomas-darling** [comment]
If injecting the router-view element is not possible, how about registering a `Promise` for it instead, thus allowing that to be injected instead? We don't really need to access the element in the constructor, so if we could just have a promise that is resolved with the router-view element once routing is complete, that would be just as good - and I think this is something that really needs to be supported in one way or another :-)

**Comment by thomas-darling** [comment]
Actually, it would also be fine to just register the element whenever it becomes available. Sure, it won't work with constructor injection then, but at least it is then _possible_ to get it, by injecting the container and then manually resolving it later using `container.get(Element)`. 

I'm sure there's a better way to do this, but playing around with it, this hack does seem to work:
```js
const _process = RouterView.prototype.process;
RouterView.prototype.process = function (viewPortInstruction) {
  viewPortInstruction.component.childContainer.registerInstance(Element, this.element);
  return _process.apply(this, arguments);
};
```

Now, for my particular use case, this is only a partial solution, as attributes applied to the `<template>` for a router view are still not copied to the actual router-view element - but maybe that could be supported?
In that case, I can create a separate issue for that :-)

**Comment by EisenbergEffect** [comment]
I'd be willing to take a PR to register a special service that locates the router-view. For example, in the router library we might have:

```javascript
export class ViewPortLocator {
  findNearestViewPort(): Promise
  _notifyOfNearestViewPort(routerView);
}
```

This could be registered up front and your independent service could take a dependency on it. The promise would resolve when the router-view is available. Thoughts?

**Comment by thomas-darling** [comment]
That would work perfectly, and would be more testable, compared to just registering and manually resolving the element after the construction of the view model - same reason I initially suggested registering a promise, but of course, with that we would have the problem of which key to use.
I think your suggestion is the way to go - I'll prepare a pull request, probably later today :-)

**Comment by thomas-darling** [comment]
Reading through the router code, it looks like both the child container and the component view model is instantiated by the `TemplatingRouteLoader` in the `templating-router` repository, and the `RouterView` also resides in there - so I'm thinking this service should probably also live in that repository. Agree?

**Comment by EisenbergEffect** [comment]
We might want to put an abstract base class for this in the router repo, with the implementation being specific to the templating-router repo.

**Comment by thomas-darling** [comment]
We could definitely do that - then the concrete implementation would just be registered using the base class as key, exactly like the route loader.

Whether we need it probably depends on where it would be most appropriate to call `_notifyOfNearestViewPort()`. We could _technically_ do it from inside the `process` method of the `RouterView` - in that case, I don't think we really need a base class, unless there are other reasons I'm not aware of?

I'm not really sure where it would be best to place this call though - inside the `process` method does feel a little strange, but digging around in the code, it looks like the alternative would be just _before_  the call to `process`, at https://github.com/aurelia/router/blob/master/src/navigation-instruction.js#L207
That seems to be the earliest point where aurelia currently attempts to locate the router view.
If you think we should call it earlier, I'm afraid I'll need a pointer to where that might be - there's a lot going on in there :-)

Oh, and another thing, what would you prefer `findNearestViewPort()` to return a promise of?
What we need is the DOM element, so I'd suggest `Promise<Element>`, but is this what you were suggesting, or do you prefer `Promise<RouterView>`? If you prefer `RouterView`, then I think it would be best to keep the two in the same repository. I also just noticed that `element` is currently not included in the typings for `RouterView`, so we would also need to add that.

**Comment by EisenbergEffect** [comment]
All good questions. It's probably nice to return `Promise<RouterView>` and then add an `element` property onto that type so that consumers can access either. Since the implementation will be RouterView specific, I think we should probably have the code inside the `process` method rather than inside `navigation-instruction`. In light of that, we probably can't have an abstraction inside the `router` repo, since the types won't be available. I think that's probably ok, since this is really a detail that's very specific to `templating-router`. That happens to be the only real implementation so, I think we can put it there and avoid any unnecessary "hoops".

**Comment by thomas-darling** [comment]
Sounds good - I'll have a pull request ready to go shortly :-)
Do you still prefer `ViewPortLocator`, or would `RouterViewLocator` be a better name?

**Comment by EisenbergEffect** [comment]
Since it will be specific to the templating-router implementation, I think we cdan name it RouterViewLocator. What do you think?

**Comment by thomas-darling** [comment]
Ok, so this is actually an interesting question, now that I think more about it.

As we discussed this, I somehow came to look at this from the viewpoint of "how to locate the nearest router view", despite my own use case, and I suspect maybe also your initial reply, actually being more along the lines of "how to locate the nearest _element_".

If we keep this focused on just locating the `RouterView`, then I'm in favor of the name  `RouterViewLocator`.
I actually have an implementation of that ready to go already.

**The alternative**
We _could_ also keep the name `ViewPortLocator`, and actually have it locate the nearest _view port_ - i.e. the nearest _element_, which would be either the element registered in the component container, or if that is not found, the router view element, which would then be located as we've talked about here.
If we go down that path instead, then it suddenly makes more sense to resolve the promise to an `Element`, and maybe also to have a base class in the `router` repository - although in that case, it would actually start having as much to do with templating as routing.

Despite my own use case here, I'm leaning towards keeping this focused on just the router view though - i.e. to stick with `RouterViewLocator`. Otherwise I think the `templating-router` library starts being a little less coherent, with assumptions about custom elements having their `Element` registered in child containers, etc.
What's more, once we have a `RouterViewLocator` it should be trivial to build a `ViewPortLocator` using it - it would just need to walk up the container hierarchy, until it finds a registration for either an `Element` or a `RouterViewLocator`, and then use that.

I guess the real question here is, whether something like this might be needed in other parts of aurelia, and if so, which package references are allowed?

Right now, the only use case I can think of inside of aurelia, is something I mentioned earlier:
I'd like to enable attributes from `<template>` elements for router views to be copied to the actual `<router-view>` element, similar to how it's done for custom elements.
The code for that is in the `templating` repository (https://github.com/aurelia/templating/blob/a1fcdffbcb1b0dcfbd23387c467302d540a9144f/src/view-factory.js#L136), and while I haven't yet looked too deeply into how feasible this feature might be, an implementation would need to somehow get a reference to the router view element...

What's your take on this?
Should we keep this focused on just the `router-view`, or do you think that would become a limitation?
If we keep it as just the `RouterViewLocator`, then I've got a pull request ready to go :-)

**Comment by EisenbergEffect** [comment]
Let's keep it as RouterViewLocator for now. It would be easy to build a service for element location on top of this by composing the RouterViewLocator, subscribing to it and then publishing out the element associated with the RouterView.

**Comment by thomas-darling** [comment]
Sounds good to me - I've created a pull request for the feature.
Let me know if there's any adjustments needed :-)
